src/cocotb/handle.py

```diff
@@ -316,30 +472,50 @@ def _sub_handle_key(self, name):
         else:
             raise ValueError(f"Unable to match an index pattern: {name}")
 
-    @lru_cache(maxsize=None)
-    def __len__(self) -> int:
-        self._discover_all()
-        return len(self._sub_handles)
+    def _child_path(self, key: int) -> str:
+        return f"{self._path}[{key}]"
 
-    def __getitem__(self, index):
-        if isinstance(index, slice):
-            raise IndexError("Slice indexing is not supported")
-        if index in self._sub_handles:
-            return self._sub_handles[index]
-        new_handle = self._handle.get_handle_by_index(index)
-        if not new_handle:
-            raise IndexError("%s contains no object at index %d" % (self._name, index))
-        path = self._path + "[" + str(index) + "]"
-        self._sub_handles[index] = SimHandle(new_handle, path)
-        return self._sub_handles[index]
+    def _get_handle_by_key(self, key: int) -> Optional[simulator.gpi_sim_hdl]:
+        return self._handle.get_handle_by_index(key)
+
+    def __getitem__(self, key: int) -> SimHandleBase:
+        if isinstance(key, slice):
+            raise TypeError("Slice indexing is not supported")
+        try:
+            return super().__getitem__(key)
+        except KeyError as e:
+            raise IndexError(str(e)) from None
+
+    @cached_property
+    def range(self) -> Range:
+        """Return a :class:`~cocotb.types.Range` over the indexes of the array/vector."""
+        left, right = self._handle.get_range()
+
+        # guess direction based on length until we can get that from the GPI
+        length = self._handle.get_num_elems()
+        if length == 0:
+            direction = "to" if left < right else "downto"
+        else:
+            direction = "to" if left <= right else "downto"
+
+        return Range(left, direction, right)
+
+    def left(self) -> int:
+        """Return the leftmost index in the array/vector."""
+        return self.range.left
+
+    def direction(self) -> str:
+        """Return the direction ("to"/"downto") of indexes in the array/vector."""
```

Comment:
```suggestion
        """Return the direction (``"to"``/``"downto"``) of indexes in the array/vector."""
```
Prevent smartquotes.